### DISCUSSION TITLE
SQLAlchemy 1.4 - IntegrityError: raised as a result of Query-invoked autoflush

### DISCUSSION THREAD

**Discussion by sh4nks** [sqlalchemy/sqlalchemy#6766] [Usage Questions] [answered]
Hi, I have been trying to get SQLAlchemy 1.4 to work with flaskbb but I am always running into the same problem:
```
sqlalchemy.exc.IntegrityError: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely)
```
It works without problems with the latest 1.3 release.

A minimal working example can be found here: https://github.com/sh4nks/sa-14-mcve
Any ideas why it's not working with 1.4?

I'd really appreciate any help or input!

**Comment by zzzeek** [comment, marked as answer]
hi - 

the stack trace shows where the problem is:

  File "/home/peter/Development/flaskbb-sa14/app.py", line 497, in create_test_data
    topic.save(post=post, user=user1, forum=forum)
  File "/home/peter/Development/flaskbb-sa14/app.py", line 277, in save
    self.username = user.username

"user.username" is attempting to load data.   this triggers an autoflush which attempts to flush the "topic" object above.   This then fails because the "username" field is empty.

"topic" is present in the session here, which perhaps wasn't the case before.  I can't go further than that since the line numbers in the posted stack trace don't line up with what's in the source code for app.py.   However, use the [no_autoflush](https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=no_autoflush#sqlalchemy.orm.Session.no_autoflush) context manager to prevent the flush within this block.     this issue is typically caused by the [legacy backref behavior](https://docs.sqlalchemy.org/en/14/orm/cascades.html#controlling-cascade-on-backrefs) which is probably whats happening here, so you might want to use a Session with future=True and/or the [cascade_backrefs](https://docs.sqlalchemy.org/en/14/orm/relationship_api.html#sqlalchemy.orm.relationship.params.cascade_backrefs) flag to prevent "topic" from being added to the session prematurely.

**Reply by sh4nks** [reply]
Hi, thanks for your reply! Setting future=True solves this problem. I still have a couple of other errors but I believe I can solve them with the input you have provided.
